File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxPathSum(TreeNode root) {
        int[] ans = new int[1];
        ans[0] = root.val;
        helper(root,ans);

        return ans[0];

    }

    int helper(TreeNode root, int[] ans){

        if (root == null){
            return 0;
        }

        int left = Math.max(0,helper(root.left,ans));
        int right = Math.max(0,helper(root.right,ans));

        ans[0] = Math.max(ans[0],left + right + root.val);
        ans[0] = Math.max(ans[0], left + root.val);
        ans[0]  = Math.max(ans[0],right+root.val);

        return root.val + Math.max(left,right);


    }
}

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        ans = [-math.inf]

        def helper(root, ans):

            if root is None:
                return 0
            
            l = helper(root.left, ans)
            r = helper(root.right, ans)

            ans[0] = max(ans[0], root.val, root.val + l, root.val + r, root.val + l + r)

            return max(0, root.val, root.val + max(l,r))
        
        helper(root, ans)
        return ans[0]

            